### PR TITLE
cp

### PR DESCRIPTION
**Pull request by daveey**

### CODE REVIEW

**Comment by daveey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1400?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1400** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1400?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1400?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1398** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1398?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by relh — APPROVED** [review]
This is great! 

There are two paradigms I think we could use:

1) Our code is written to basically only know about one-gpu worth of stuff, and then we multiple by world size for logging and/or choose values like learning rate based off of it (this fits with having only _master log as well)

2) Our code knows about the size of the world and understands the global batch size and divides things up as needed.

Having done tons of `torch.distributed` training I really like the version where the training code is written mostly for one-gpu mental model with a few carefully placed barriers for syncing. Right now we have quite a few _master checks for things like loading/saving/logging etc. I think we should think about training code later 

One nice thing is that I almost have never found a situation where torch.distributed breaks from going from one node to many nodes. Basically if torch.distributed works for going from 1 GPU to 4 GPUs on the same machine it works beyond that. There used to be two packages, one for single machine multiple gpus and one for multiple machines but then torch.distributed (the multiple machine one) just won out because it worked so well
